api/src/main/java/org/apache/iceberg/expressions/UnboundPredicate.java

```diff
@@ -125,13 +180,65 @@ public Expression bind(Types.StructType struct, boolean caseSensitive) {
         case LT_EQ:
         case EQ:
           return Expressions.alwaysFalse();
-//        case IN:
-//          break;
-//        case NOT_IN:
-//          break;
       }
     }
     return new BoundPredicate<>(op(), new BoundReference<>(field.fieldId(),
-        schema.accessorForField(field.fieldId())), lit);
+            schema.accessorForField(field.fieldId())), lit);
+  }
+
+  @SuppressWarnings("unchecked")
+  private Expression bindInOperation(Types.NestedField field, Schema schema) {
+    final Set<Literal<T>> lits = literals().stream().map(
+        l -> {
+          Literal<T> lit = l.to(field.type());
+          if (lit == null) {
+            throw new ValidationException(String.format(
+                "Invalid value for comparison inclusive type %s: %s (%s)",
+                field.type(), l.value(), l.value().getClass().getName()));
+          }
+          return lit;
+        })
+        .filter(l -> l != Literals.aboveMax() && l != Literals.belowMin())
+        .collect(Collectors.toSet());
+
+    if (lits.isEmpty()) {
+      return Expressions.alwaysFalse();
+    } else if (lits.size() == 1) {
+      return new BoundPredicate<>(Operation.EQ, new BoundReference<>(field.fieldId(),
+          schema.accessorForField(field.fieldId())), lits.iterator().next());
+    } else {
+      return new BoundSetPredicate<>(Operation.IN, new BoundReference<>(field.fieldId(),
+          schema.accessorForField(field.fieldId())), lits);
+    }
+  }
+
+  @Override
+  public String toString() {
```

Comment:
This could be implemented in `Predicate` by adding an abstract protected method `literalString` that converts a literal or literalSet to a string.

Reply:
👌